[wip] QAT ConvBN: remove explicit folding and use BN instead

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38478 [wip] QAT ConvBN: remove explicit folding and use BN instead**

Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Training MobileNetV2 on Imagenet-1k with QAT - accuracy curves match exactly before and after this diff


Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21603230](https://our.internmc.facebook.com/intern/diff/D21603230)

vkuzo committed:
[wip] QAT ConvBN: remove explicit folding and use BN instead

Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_qat_module.py:
```diff
@@ -82,6 +270,7 @@ def test_conv_bn_relu(
 
             cls = ConvBnReLU2d if use_relu else ConvBn2d
             qat_op = cls(
+                copy.deepcopy(bn_op),
```
can you move bn after the conv2d parameters

vkuzo replied:
one limitation is that some of the conv params are optional, and all required params have to come before optional params.  I thought this was worth changing the arg order (so we can make bn required), but flexible - lmk.

jerryzh168 replied:
OK

vkuzo replied:
actually, I take it back, passing the module instead of params leads to other problems.  Will switch back to params and move it after conv

raghuramank100 commented on torch/nn/intrinsic/qat/modules/conv_fused.py:
```diff
@@ -49,83 +44,45 @@ def __init__(self,
+        running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
+        scale_factor = self.bn.weight / running_std
         scaled_weight = self.weight * scale_factor.reshape([-1, 1, 1, 1])
+        # this does not include the conv bias
```
Defining scaled_weight as:
```
 scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape([-1, 1, 1, 1]))/scale_factor.reshape([1, -1, ,1,1])
```
simplifies the code to:
```
 conv = self._conv_forward(input, scaled_weight)
```

raghuramank100 reviewed (approved):
Looks good to me!

jerryzh168 commented on test/quantization/test_qat_module.py:
```diff
@@ -3,19 +3,207 @@
 import torch.backends.mkldnn
 from torch.testing._internal.common_utils import TestCase
-from hypothesis import given
+from hypothesis import given, seed
```
lint

jerryzh168 reviewed (approved):
Looks good

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[wip] QAT ConvBN: remove explicit folding and use BN instead"


Summary:

Before this PR, the QAT ConvBN module inlined the batch normalization code
in order to reproduce Conv+BN folding.

This PR updates the module to use BN directly.  This is mathematically
equivalent to previous behavior as long as we properly scale
and fake quant the conv weights, but allows us to reuse the BN code
instead of reimplementing it.

In particular, this should help with speed since we can use dedicated
BN kernels, and also with DDP since we can hook up SyncBatchNorm.

Test Plan:

```
python test/test_quantization.py TestQATModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21603230](https://our.internmc.facebook.com/intern/diff/D21603230)

[ghstack-poisoned]
